### ISSUE TITLE
How to resize columns automatically?

### ISSUE THREAD

**Issue by DGOIT-Jan** [rabanti-github/NanoXLSX#18]
Hi Raphael,

thanks for this little library. Just found it and it works very well!

One nice enhancement would be the auto-resizing of the columns based on the contents of the cells (also taking the auto-filter dropdown-button into account)!

Could you take a look into this, please?

Thanks and regards,
Jan

**Comment by rabanti-github** [comment]
Hello

Thank you for your question.
Automatic column or row resizing is at the moment not possible.
The reason for this is that auto widths or heights are dependent on font metrics. Especially tricky with inline formatting of Text in cells (which is not implemented yet).
As I cannot really say what fonts are installed on the target system, a prediction of these metrics is difficult.
Furthermore, when it comes to formulas, the value of the cell formula must be completely resolved (which is not implemented yet) to evaluate the width of the cell.

Currently, I am on a big update (on dev branch). Maybe I can have a look into the topic afterwards. Inline formatting of cells is on the backlog anyway.

**Comment by DGOIT-Jan** [comment]
Hi Raphael,

thanks for your quick reply.

Just found two more things that may not work as intended (or I'm doing something wrong :))

I try to hide the first sheet with:

`_workbookNanoXlsx.Worksheets.First().Hidden = true;`

The property is then set to true but in the saved file, the sheet is still visible.

And I thought I could use the SheetId as an identifier with my own numbers, but internally, the id is just counted up.

I.e. `var ws = new Worksheet($"{hrUser.Surname} ({hrUser.EmployeeId})", hrUser.EmployeeId, _workbookNanoXlsx);`

The "hrUser.EmployeeId" is 12345 but the SheetId remains 1 and so on...

If one can't use the Id it should be removed from the constructor and from the public properties also.

Regards

**Comment by rabanti-github** [comment]
Hi
The first one should actually work (the property usage). I have to look into this. Are you sure the LINQ operation does not return just a copy (object) of the worksheet, while the original one is still in the unaltered state?

The second one:
The internal number should better not be used, since it is important for the creation of some references. maybe I have to set this to internal in the future. However, you can use the worksheet name to identify your sheets.
The upcoming update will make it possible to select the worksheet directly by the name and not just the index.

Please let me know, If you find any other issues, since I have the opportunity to fix them with the pending update.

Thanks and greetings

**Comment by DGOIT-Jan** [comment]
The hiding works on all but the first worksheet.

It doenst matter if I set it via 
`_workbookNanoXlsx.Worksheets.First().Hidden = true;`
or 
`_workbookNanoXlsx = new NanoXLSX.Workbook(@"d:\temp\test.xlsx","Info");`
`_workbookNanoXlsx.CurrentWorksheet.Hidden = true;`

This works as expected:
`_workbookNanoXlsx.Worksheets.Skip(1).First().Hidden = true;`
and this works, too:
`var ws = new Worksheet($"{hrUser.Surname} ({hrUser.EmployeeId})", hrUser.EmployeeId, _workbookNanoXlsx)
                {
                    Hidden = true
                };`

Just a note: If I try to hide ALL sheets, the resulting file seems damaged:

![image](https://user-images.githubusercontent.com/79843347/132681042-1dcc1e83-94b0-43be-9d37-23accb9feec5.png)

So, Excel seems to be a bit confused and don't know what to show if all sheets are hidden ;)

However, if you try to recover the file, the first sheet is still visible...

**Comment by rabanti-github** [comment]
Can you attach an example of the corrupt workbook?
I will have a look into these two issues.

Thank you in advance.

**Comment by DGOIT-Jan** [comment]
Sure, here it is.

[test2.xlsx](https://github.com/rabanti-github/NanoXLSX/files/7136412/test2.xlsx)

**Comment by DGOIT-Jan** [comment]
As for the auto-sizing:
Too bad you're not able to detect the standard-font of Excel but if so, you might be able to implement this feature just for this font as a start.

I know its possible using Office Interop but thats not an option here...

Or you could just implement it for the default font, which should be Calibri, 11.

**Comment by rabanti-github** [comment]
Well, technically it seems not to be possible that all worksheets are hidden in a workbook. Excel complains when you try that. So, the possibility of hiding all worksheets in NanoXLSX is actually a bug and results in a broken xlsx.
However, Excel allows to hide all worksheets using the workbook functionality `<workbookView visibility="hidden" ... [internal parameters] />`

This meaans:
1. Hiding all worksheets has to be prevented (throw an Exception on the last one)
2. The workbook needs a new property to set the visibility state of the whole workbook (beside hiding the main window). This property overrides all Hidden properties of the worksheets and hides all

Thanks for the report!
Please give me some days for that. I cannot fix this until Saturday night (MET) or Sunday

**Comment by rabanti-github** [comment]
OK, I have an update for you.
The functionality to hide all worksheets is already implemented (lol, completely forgot about that).
You can use: `workbook.Hidden = true;`
The property name may be a little bit misleading, since it implies that the whole workbook is hidden after set to true (no window visible), but actually at least an empty excel application remains open.

The issue with the worksheet with index 0 that cannot be hidden is still under investigation

**Comment by rabanti-github** [comment]
Hello,
The new release v1.8.1 fixes the issue that the first worksheet could not be deleted.
The NuGet package is already available.
However, there are some restrictions that I  had to introduce:
- If the selected worksheet (use workbook.SetSelectedWorksheet(int)) is set to hidden, an Exception will be thrown
- An empty workbook (achieved by removing all worksheets) will throw an exception
- If all worksheets are set to hidden, an Exception will be thrown

Regarding the issue with the column auto width: I have still some considerations to do. I will post an analysis later

**Comment by rabanti-github** [comment]
OK, here are some considerations about auto-width of columns or auto-height of rows

### TLDR;
At the moment, it is not feasible to implement a auto-size function for column widths, since there are too many uncertainties.

### Built-in functionality of Excel
Excel has an auto-fit function on the cell formatting. However, this function is not persistent. Every time we change the content of a cell (assumed: we heave only this cell with values on the column), the column width remains. Values may overlap over the column or there is a space between the text and the column boundary. There is also no option to auto-update this function (which is kind of sad/puzzling/infuriating/stupid). That said, there is also no possibility to define something like that in an Excel file, without using VBA

### Fonts
Fonts, or better yet, font metrics are a quite tricky topic. Each font has varying metrics. Not only the size of the font matters, but the letter spacing and the particular width of the characters. An ‘I’ has obviously a completely different width than a ‘W’. An exception are of course mono-space fonts. These are usually not very good when it comes to the readability of structured text (important for programming, but not for novels etc.). Of course, “width tables” could be crated to predict the particular spacings of characters. However, these tables would be different for every font.
Another problem is that it’s not possible to know about the installed fonts on the systems where Excel files are viewed. OK, one could assume that “Calibri” is available on every system. But it remains an assumption that could change with any Windows update (e.g. Windows 11)

### Formulas
An extremely complex problems are formulas. It is almost impossible to predict the output of formulas without knowing about all inputs and 100% accurate parsing of the formulas. Especially problematic, when a reference to a cell points to another reference, and so on. And even we would resolve the whole chain of references, links to external files break any resolution. That said, even simple formula like “=B2&B3” could lead to any value from empty to hundreds of characters.

### Localization
Another problem could be localization of computer systems. So, if we compare for instance the default date format of EN-US “9/18/2021” with DE-DE “18.09.2021”, we can already see a difference in the number of characters. A date, time or even numbers will certainly have various lengths on systems with different cultures.

### Number formatting
Number Formatting, especially custom formatting, has a huge effect on the display value of a cell. One of the most extreme examples is style format number 14 or 21. If set, a number becomes a date (14) or time (21). And when it comes to custom formats, the complexity of the formatting can be almost infinite. Similar to the formulas, formatting would have completely to be parsed to determine the number of characters of a cell.

### Inline cell formatting
Although not yet defined in NanoXLSX, the inline formatting of cells can lead to completely different widths. Let’s say, a character in a text has a different size or even font, we already have a problem with the font metrics, since they are mixed within the cell.

### Conclusion
If we ignore the missing persistent function in Excel, we still have four to five levels of uncertainty to calculate the width of a column. The implement such a function would make it necessary to introduce several, complex parsers, lookup tables and assumptions. The effort to do that is at the moment not feasible, since it would be more powerful than the current column implementation of Excel (although it has all the parsers and knowledge about installed fonts). And even the naïve implementation of a function that considers Calibri, size 11, without formulas or formatting, would be probably not satisfying.

**Comment by DGOIT-Jan** [comment]
Hi Raphael,
sorry for my late reply, have been very busy during the last weeks...

Thanks for your effort and your thoughts about this topic!

In Excel, I just click the upper left "box" of the sheet, so all content is selected. After that, I just double-click a splitter (i.e. between column A and B) and the columns are resized nicely...

Best regards!

**Comment by rabanti-github** [comment]
Hi
I am aware of this function in Excel. What I actually  wanted to point out is that the column width is not dynamically resizing when the content of a cell changes. You always has to apply this function (as far as I know from Office365).
However, my comment is not meant as a hard 'won't fix'. As I am planning to modularize NanoXLSX, an optional package to implement a Font-Helper could be realized.
I know exactly the pain of column widths. I am with you :-). Hell, I am hate it to resize columns based on heuristics. But at the moment, a reliable solution would be too time consuming and would take too much place in the core library.

What I could propose at the moment to you:
- Create a worksheet in Excel with a quite big number of columns
- Each column contains a random String with a random number of characters
- Resize the the columns in Excel and write down the number of characters, as well as the column width, determined by Excel
- Create a statistical point chart where X is the number of characters, and Y ist the column length (or vice versa)
- After you have a number of points (should display an ascending point cloud), you can add an extrapolation function (probably linear). With this function you can calculate the (approximate) average column width of a column with a certain number of characters.

However, this method will not save you from the problem of formatting or dynamic cell content, by formulas.

Is it OK for you if I close the issue?
Please let me know if you need further assistance.

Greetings,
Raphael

**Comment by rabanti-github** [comment]
I will close this issue for now.
Please do not hesitate to ask further questions, to file bugs or to propose features.
The whole column with thing is right behind in-line formatting of cells on my roadmap.
I will most likely have a look at this topic, after I released - the currently in development - v2.0
